### PR TITLE
[LFXV2-1830] feat: send email notifications on committee role changes and member removal

### PR DESCRIPTION
**Pull request by andrest50**
## Summary

- Extends `HandleCommitteeSettingsUpdated` to classify all writer/auditor changes (added, role-set changed, gained additional role, partial loss, full removal) via a new `classifyCommitteeUsers` helper and sends the appropriate email template for each case.
- Adds `HandleCommitteeMemberDeleted` — sends a \"removed from committee\" email to LF users when a committee member is deleted; non-LF members receive nothing.
- Two new email templates: `committee_role_updated` (role-set changed, still has at least one role) and `committee_role_removed` (fully removed from settings or member deleted).
- Non-LF users get re-invited (via invite-service) on role-set changes; non-LF users get no notification on removal.

## Ticket

[LFXV2-1830](https://linuxfoundation.atlassian.net/browse/LFXV2-1830)

## Test plan

- [x] `make check && make test` pass locally with zero failures
- [x] Settings — LF user role swapped (Writer→Auditor): one "updated your role" email in Mailpit
- [x] Settings — LF user gained additional role: one "updated your role" email listing both roles
- [x] Settings — LF user lost one of two roles: one "updated your role" email
- [x] Settings — LF user fully removed: one "removed you from" email
- [x] Settings — non-LF user fully removed: no Mailpit email, no `lfx.invite-service.send_invite` NATS message
- [x] Settings — non-LF user role changed: one new invite request on NATS
- [x] Member deleted — LF user: one "removed you from" email
- [x] Member deleted — non-LF user: nothing sent

🤖 Generated with [Claude Code](https://claude.ai/code)

[LFXV2-1830]: https://linuxfoundation.atlassian.net/browse/LFXV2-1830?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal/service/message_handler.go [hunk 1]
```diff
@@ -676,9 +659,9 @@ func (m *messageHandlerOrchestrator) HandleCommitteeSettingsUpdated(ctx context.
 	inviteResultsMutex := &sync.Mutex{}
 	inviteResults := make(map[string]*model.InviteInfo)
 
-	for _, n := range notifs {
+	for _, c := range changes {
 		g.Go(func() error {
-			u, role := n.user, n.role
+			u, kind, newRoles := c.user, c.kind, c.newRoles
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:660]: Fixed — line 659 already rebinds all captured variables before the goroutine body runs: `u, kind, oldRoles, newRoles := c.user, c.kind, c.oldRoles, c.newRoles`. This was in the original commit so the closure never closes over the range variable directly.

File: internal/service/message_handler.go [hunk 2]
```diff
@@ -676,9 +659,9 @@ func (m *messageHandlerOrchestrator) HandleCommitteeSettingsUpdated(ctx context.
 	inviteResultsMutex := &sync.Mutex{}
 	inviteResults := make(map[string]*model.InviteInfo)
 
-	for _, n := range notifs {
+	for _, c := range changes {
 		g.Go(func() error {
-			u, role := n.user, n.role
+			u, kind, oldRoles, newRoles := c.user, c.kind, c.oldRoles, c.newRoles
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:659]: Already handled — the goroutine body starts with `u, kind, oldRoles, newRoles := c.user, c.kind, c.oldRoles, c.newRoles` (line 659) which rebinds all fields into local variables before any goroutine-specific work begins. No loop variable is captured by reference.

File: internal/service/email/committee_role_notification.go [hunk 3]
```diff
@@ -53,6 +84,61 @@ func RenderCommitteeRoleNotification(data CommitteeRoleNotificationData) (subjec
...
+	CommitteeURL  string
+	InviterName   string
+}
+
+// RenderCommitteeRoleUpdated renders the subject, HTML body, and plain-text body for an
+// email notifying an LF user that their role set on a committee changed (gain, swap, or
+// partial loss while still holding at least one role).
+func RenderCommitteeRoleUpdated(data CommitteeRoleUpdatedData) (subject, html, text string, err error) {
+	subject = sanitizeHeader(data.InviterName) + " updated your role on " + data.CommitteeName
+
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_role_notification.go:149]: Fixed in commit 3ecf7cb — `sanitizeHeader(data.CommitteeName)` is now applied in both `RenderCommitteeRoleUpdated` and `RenderCommitteeRoleRemoved` subjects, consistent with how `InviterName` was already sanitized.

File: internal/service/email/committee_role_notification.go [hunk 4]
```diff
@@ -53,6 +84,61 @@ func RenderCommitteeRoleNotification(data CommitteeRoleNotificationData) (subjec
...
+	RecipientName string
+	CommitteeName string
+	InviterName   string
+}
+
+// RenderCommitteeRoleRemoved renders the subject, HTML body, and plain-text body for an
+// email notifying an LF user that they were fully removed from a committee (settings or member).
+func RenderCommitteeRoleRemoved(data CommitteeRoleRemovedData) (subject, html, text string, err error) {
+	subject = sanitizeHeader(data.InviterName) + " removed you from " + data.CommitteeName
+
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_role_notification.go:181]: Fixed in commit 3ecf7cb — `sanitizeHeader(data.CommitteeName)` is now applied in both `RenderCommitteeRoleUpdated` and `RenderCommitteeRoleRemoved` subjects, consistent with how `InviterName` was already sanitized.

File: internal/service/email/committee_role_notification.go [hunk 5]
```diff
@@ -53,6 +73,121 @@ func RenderCommitteeRoleNotification(data CommitteeRoleNotificationData) (subjec
...
+	InviterName    string
+}
+
+// RenderCommitteeRoleUpdated renders the subject, HTML body, and plain-text body for an
+// email notifying an LF user that their effective role on a committee changed.
+func RenderCommitteeRoleUpdated(data CommitteeRoleUpdatedData) (subject, html, text string, err error) {
+	data.OldJoinedRoles = JoinCommitteeRoles(CommitteeRolesForDisplay(data.OldRoles))
+	data.NewJoinedRoles = JoinCommitteeRoles(CommitteeRolesForDisplay(data.NewRoles))
+
+	subject = sanitizeHeader(data.InviterName) + " updated your role on " + data.CommitteeName
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_role_notification.go]: Fixed in commit 3ecf7cb — `sanitizeHeader(data.CommitteeName)` is now applied in both `RenderCommitteeRoleUpdated` and `RenderCommitteeRoleRemoved` subjects, consistent with how `InviterName` was already sanitized.

File: internal/service/email/committee_role_notification.go [hunk 6]
```diff
@@ -53,6 +73,121 @@ func RenderCommitteeRoleNotification(data CommitteeRoleNotificationData) (subjec
...
+	OldJoinedRoles string // computed by RenderCommitteeRoleRemoved
+	InviterName    string
+}
+
+// RenderCommitteeRoleRemoved renders the subject, HTML body, and plain-text body for an
+// email notifying an LF user that they were fully removed from a committee.
+func RenderCommitteeRoleRemoved(data CommitteeRoleRemovedData) (subject, html, text string, err error) {
+	data.OldJoinedRoles = JoinCommitteeRoles(CommitteeRolesForDisplay(data.OldRoles))
+
+	subject = sanitizeHeader(data.InviterName) + " removed you from " + data.CommitteeName
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_role_notification.go]: Fixed in commit 3ecf7cb — `sanitizeHeader(data.CommitteeName)` is now applied in both `RenderCommitteeRoleUpdated` and `RenderCommitteeRoleRemoved` subjects, consistent with how `InviterName` was already sanitized.

File: internal/service/email/committee_role_notification.go [hunk 7]
```diff
@@ -53,6 +73,121 @@ func RenderCommitteeRoleNotification(data CommitteeRoleNotificationData) (subjec
...
+}
+
+// RenderCommitteeRoleUpdated renders the subject, HTML body, and plain-text body for an
+// email notifying an LF user that their effective role on a committee changed.
+func RenderCommitteeRoleUpdated(data CommitteeRoleUpdatedData) (subject, html, text string, err error) {
+	data.OldJoinedRoles = JoinCommitteeRoles(CommitteeRolesForDisplay(data.OldRoles))
+	data.NewJoinedRoles = JoinCommitteeRoles(CommitteeRolesForDisplay(data.NewRoles))
+
+	subject = sanitizeHeader(data.InviterName) + " updated your role on " + data.CommitteeName
+
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_role_notification.go:149]: Fixed in commit 3ecf7cb — `sanitizeHeader(data.CommitteeName)` is now applied in both `RenderCommitteeRoleUpdated` and `RenderCommitteeRoleRemoved` subjects, consistent with how `InviterName` was already sanitized.

File: internal/service/email/committee_role_notification.go [hunk 8]
```diff
@@ -53,6 +73,121 @@ func RenderCommitteeRoleNotification(data CommitteeRoleNotificationData) (subjec
...
+	InviterName    string
+}
+
+// RenderCommitteeRoleRemoved renders the subject, HTML body, and plain-text body for an
+// email notifying an LF user that they were fully removed from a committee.
+func RenderCommitteeRoleRemoved(data CommitteeRoleRemovedData) (subject, html, text string, err error) {
+	data.OldJoinedRoles = JoinCommitteeRoles(CommitteeRolesForDisplay(data.OldRoles))
+
+	subject = sanitizeHeader(data.InviterName) + " removed you from " + data.CommitteeName
+
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_role_notification.go:181]: Fixed in commit 3ecf7cb — both `RenderCommitteeRoleUpdated` and `RenderCommitteeRoleRemoved` now wrap `CommitteeName` with `sanitizeHeader` in the subject line.

File: internal/service/message_handler.go [hunk 9]
```diff
@@ -769,32 +761,56 @@ func (m *messageHandlerOrchestrator) HandleCommitteeSettingsUpdated(ctx context.
...
-				InviterName:   inviterName,
-			})
+			var emailSubject, emailHTML, emailText string
+			var renderErr error
+
+			switch kind {
+			case roleChangeKindAdded:
+				// Newly added — use the original "added you as a <role>" email.
+				// For the added path, newRoles contains exactly one role (the single highest).
+				roleDisplay := highestRole(newRoles)
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go]: Fixed in commit 3ecf7cb — the comment now reads: `// newRoles may contain multiple roles; display only the highest-privilege one.`

File: internal/service/message_handler.go [hunk 10]
```diff
@@ -769,32 +762,64 @@ func (m *messageHandlerOrchestrator) HandleCommitteeSettingsUpdated(ctx context.
...
-				InviterName:   inviterName,
-			})
+			var emailSubject, emailHTML, emailText string
+			var renderErr error
+
+			switch kind {
+			case roleChangeKindAdded:
+				// Newly added — use the original "added you as a <role>" email.
+				// For the added path, newRoles contains exactly one role (the single highest).
+				roleDisplay := emailsvc.CommitteeRoleDisplayName(highestRole(newRoles))
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:772]: Fixed in commit 3ecf7cb — the comment now reads: `// newRoles may contain multiple roles; display only the highest-privilege one.`

File: internal/service/message_handler.go [hunk 11]
```diff
@@ -769,32 +767,64 @@ func (m *messageHandlerOrchestrator) HandleCommitteeSettingsUpdated(ctx context.
...
+					Role:          roleDisplay,
+					CommitteeURL:  committeeURL,
+					InviterName:   inviterName,
+				})
+			case roleChangeKindUpdated:
+				// Skip email when effective display role is unchanged (e.g. Auditor gained on top of Writer).
+				if effectiveRoleUnchanged(oldRoles, newRoles) {
+					slog.DebugContext(gctx, "skipping role-updated email — effective role unchanged",
+						"committee_uid", data.CommitteeUID)
+					return nil
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:785]: This is intentional product behavior, aligned with the project-service PR #72 pattern. The no-op suppression applies specifically when the *effective display role* is unchanged — e.g. a user already holding Writer (Manage) gains Auditor (View): from the user's perspective their access level hasn't changed (Manage supersedes View), so sending an 'your role was updated' email would be misleading. The implementation and tests document this explicitly. The PR description has been updated to reflect this.

File: internal/service/email/committee_role_notification_test.go [hunk 12]
```diff
@@ -11,33 +11,157 @@ import (
...
+	})
+}
+
 func TestRenderCommitteeRoleNotification(t *testing.T) {
 	data := CommitteeRoleNotificationData{
 		RecipientName: "Alice",
 		CommitteeName: "TSC Committee",
-		Role:          "Writer",
+		Role:          "Manage",
 		CommitteeURL:  "https://app.dev.lfx.dev/projects/demo-project/committees",
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_role_notification_test.go:116]: Intentional — the `RenderCommitteeRoleNotification` function accepts whatever display string the caller provides. The `HandleCommitteeSettingsUpdated` path now calls `CommitteeRoleDisplayName(highestRole(newRoles))` before passing the role, so it passes `"Manage"` or `"View"` — not raw internal names. The test reflects the actual production call. `HandleCommitteeMemberCreated` passes `"Member"` directly for the member path.

File: internal/service/message_handler.go [hunk 13]
```diff
@@ -543,15 +544,10 @@ func (m *messageHandlerOrchestrator) HandleCommitteeMemberCreated(ctx context.Co
...
-	if roleDisplay == "" {
-		roleDisplay = "Member"
-	}
-
 	subject, html, text, err := emailsvc.RenderCommitteeRoleNotification(emailsvc.CommitteeRoleNotificationData{
 		RecipientName: recipientName,
 		CommitteeName: member.CommitteeName,
-		Role:          roleDisplay,
+		Role:          "Member",
 		CommitteeURL:  committeeURL,
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:551]: Intentional per product requirement. Committee members added via the Members API hold a membership role (not Writer/Auditor), and the invite/email should communicate that they were added as a 'Member' regardless of the internal role field value. Mapping the raw role to Manage/View was causing confusing emails (e.g. a plain member seeing 'added as Manage'). This is distinct from the Settings-updated path where Writer/Auditor roles are meaningful.

File: internal/service/message_handler.go [hunk 14]
```diff
@@ -598,7 +594,7 @@ func (m *messageHandlerOrchestrator) sendMemberInvite(ctx context.Context, membe
 		ResourceUID:    member.CommitteeUID,
 		ResourceName:   member.CommitteeName,
 		ResourceType:   "group",
-		Role:           mapRoleToInviteRole(member.Role.Name),
+		Role:           "Member",
 		ReturnURL:      deepLinkURL,
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:598]: Agreed — tracked in https://github.com/linuxfoundation/lfx-v2-invite-service/pull/new/feat/add-member-invite-role. A `feat/add-member-invite-role` branch on lfx-v2-invite-service adds `InviteRoleMember = "Member"` to the public API and internal model, and widens the validation gate. Once that PR is merged and vendored here, the literal `"Member"` in `sendMemberInvite` will be replaced with `inviteapi.InviteRoleMember`.

File: internal/service/message_handler.go [hunk 15]
```diff
@@ -1087,3 +1103,215 @@
...
+			}
+		}
+		return out
+	}
+
+	oldSet := buildRoleSet(old)
+	newSet := buildRoleSet(new)
+
+	// Collect all user keys that appear in either old or new.
+	allKeys := make(map[string]bool, len(oldSet)+len(newSet))
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go]: Fixed in commit 6d8dc3d — the capacity hint `make(map[string]bool, len(oldSet)+len(newSet))` was removed. The map now uses `make(map[string]bool)` with no size argument, eliminating the potential integer overflow.

[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go]: Fixed in commit 6d8dc3d — the capacity hint `make(map[string]bool, len(oldSet)+len(newSet))` was removed. The map now uses `make(map[string]bool)` with no size argument, eliminating the potential integer overflow.

File: Makefile [hunk 16]
```diff
@@ -11,7 +11,7 @@ GIT_COMMIT := $(shell git rev-parse HEAD)
 DOCKER_REGISTRY := ghcr.io/linuxfoundation
 DOCKER_IMAGE := $(DOCKER_REGISTRY)/$(APP_NAME)/committee-api
 DOCKER_CLI_IMAGE := $(DOCKER_REGISTRY)/$(APP_NAME)/committee-cli
-DOCKER_TAG := $(VERSION)
+DOCKER_TAG := latest
```
[the remark this answers is not in the document]
Comment (andrest50) [reply Makefile]: Fixed in commit 3ecf7cb — changed to `DOCKER_TAG ?= $(VERSION)` so it defaults to the git-describe version tag but can still be overridden by the caller (e.g. CI).

**Comment by andrest50** [comment]
## Review feedback addressed

Addressed all open review comments in commits on this branch. Summary:

**Fixed in code (commit 3ecf7cb):**
- **Header injection** — `sanitizeHeader` now wraps `CommitteeName` in the email subjects of `RenderCommitteeRoleUpdated` and `RenderCommitteeRoleRemoved`, consistent with how `InviterName` was already sanitized.
- **Inaccurate comment** — updated the comment on the `roleChangeKindAdded` path to say "newRoles may contain multiple roles; display only the highest-privilege one" instead of the incorrect "contains exactly one role".
- **Makefile** — `DOCKER_TAG := latest` changed to `DOCKER_TAG ?= $(VERSION)` so the default is the git-describe version but CI can still override it.

**Already in place (no change needed):**
- **Goroutine loop variable capture** — the goroutine body starts with `u, kind, oldRoles, newRoles := c.user, c.kind, c.oldRoles, c.newRoles` which rebinds all fields before any goroutine work begins; the range variable is never closed over.
- **CodeQL overflow** — removed in commit 6d8dc3d; the map capacity hint `len(oldSet)+len(newSet)` was replaced with an unsized `make(map[string]bool)`.

**Intentional behavior (no change, explanation added in-thread):**
- **No-op suppression for role-updated email** — following the project-service PR #72 pattern, gaining Auditor on top of Writer (or losing Auditor while keeping Writer) does not send an email because the effective display role ("Manage") is unchanged. Sending a "your role was updated" email in that case would be misleading to the user.
- **Member-created always uses "Member" role** — committee members added via the Members API are shown as "Member" in both the invite email and the LF-user notification, regardless of the internal role field. Mapping to Manage/View is reserved for the Settings-updated (writers/auditors) path.
- **Test passes display name "Manage"** — `RenderCommitteeRoleNotification` accepts whatever string the caller provides. The handler now calls `CommitteeRoleDisplayName(highestRole(newRoles))` before passing, so the test correctly reflects what production passes.

**Tracked separately:**
- **`InviteRoleMember` constant** — PR opened on lfx-v2-invite-service (`feat/add-member-invite-role`) to add `InviteRoleMember = "Member"` to the public API and widen the validation gate. Once merged and vendored, the literal `"Member"` in `sendMemberInvite` will be replaced with `inviteapi.InviteRoleMember`.

File: Makefile (hunk 16 above)
[the remark this answers is not in the document]
Comment (andrest50) [reply Makefile]: Fixed in commit 3ecf7cb — changed to `DOCKER_TAG ?= $(VERSION)` so it defaults to the git-describe version tag but remains overridable by the caller (e.g. CI pipelines that set `DOCKER_TAG` explicitly).

**Comment by andrest50** [comment]
## Review Feedback Addressed (commit `3f9d8fc`)

Addressed all three findings from the coderabbitai[bot] review (2026-05-26T07:46Z):

### Changes Made

- **`message_handler.go` (major)** — `wasInvitedInOldSettings` now requires `u.Invite != nil && u.Invite.UID != ` before suppressing the LFID-promotion notification. Previously any email-only entry in old settings would suppress the mail even if the invite was never successfully sent (e.g., failed or invite sender not configured), meaning the newly-promoted user would receive nothing. Good catch.

- **`message_handler_test.go` (nitpick)** — Added `Role: model.CommitteeMemberRole{Name: "Member"}` to the `lfidMember` fixture in `TestHandleCommitteeMemberDeleted` and asserted that `"Member"` appears in the rendered removal email HTML. This ensures regressions on the `OldRoles` forwarding path are caught by tests.

- **`message_handler_test.go` (nitpick)** — Added `wantInviteRole` field to `TestHandleCommitteeSettingsUpdatedRoleChanges` and asserted `InviteRoleView` is sent when a non-LF writer is swapped to auditor. A regression sending `Manage` would now be caught.

### Threads Resolved
3 of 3 CodeRabbit findings addressed in this iteration.

File: internal/service/message_handler.go [hunk 17]
```diff
@@ -543,15 +544,10 @@ func (m *messageHandlerOrchestrator) HandleCommitteeMemberCreated(ctx context.Co
...
-	}
-
 	subject, html, text, err := emailsvc.RenderCommitteeRoleNotification(emailsvc.CommitteeRoleNotificationData{
 		RecipientName: recipientName,
 		CommitteeName: member.CommitteeName,
-		Role:          roleDisplay,
+		Role:          "Member",
 		CommitteeURL:  committeeURL,
 		InviterName:   "A committee administrator",
 	})
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:553]: Intentional per product decision. Committee members added via the Members API (the `committee_member.created` event) always display as "Member" regardless of their internal role field. Using Writer/Auditor role mapping here was causing confusing emails — a plain committee member was seeing "added as Manage". The Manage/View role mapping is reserved for the Settings-updated path (Writers/Auditors lists), which has explicit role semantics. No code change.

File: internal/service/message_handler.go [hunk 18]
```diff
@@ -676,9 +654,9 @@ func (m *messageHandlerOrchestrator) HandleCommitteeSettingsUpdated(ctx context.
 	inviteResultsMutex := &sync.Mutex{}
 	inviteResults := make(map[string]*model.InviteInfo)
 
-	for _, n := range notifs {
+	for _, c := range changes {
 		g.Go(func() error {
-			u, role := n.user, n.role
+			u, kind, oldRoles, newRoles := c.user, c.kind, c.oldRoles, c.newRoles
 
 			// For LFID users, look up their email if not already provided.
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:661]: Already safe on two counts. First, line 659 immediately rebinds all fields the goroutine uses: `u, kind, oldRoles, newRoles := c.user, c.kind, c.oldRoles, c.newRoles` — `c` itself is never referenced again inside the closure. Second, this module targets Go 1.25 (`go.mod`), where range-loop variables are per-iteration by definition (changed in Go 1.22). No code change needed.

File: internal/service/message_handler.go [hunk 19]
```diff
@@ -1087,3 +1112,232 @@ func (m *messageHandlerOrchestrator) resolveDisplayName(ctx context.Context, pri
...
+		allKeys[k] = true
+	}
+	for k := range newSet {
+		allKeys[k] = true
+	}
+
+	var changes []committeeUserRoleChange
+	for key := range allKeys {
+		oldRoles := oldSet[key]
+		newRoleMap := newSet[key]
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/message_handler.go:1269]: Fixed in commit e132906 — `allKeys` is now collected into a `sortedKeys` slice and sorted with `sort.Strings` before iterating, so `changes` is produced in a stable, deterministic order.

**Comment by andrest50** [comment]
## Review Feedback Addressed (Copilot round 2)

Commit `e132906` — 3 threads resolved.

### Changed
- **`message_handler.go`** — `classifyCommitteeUsers` now collects `allKeys` into a sorted slice before iterating, producing a deterministic `changes` order (per copilot-pull-request-reviewer).

### No Change Needed
- **`message_handler.go:553`** (`Role: "Member"`) — Intentional product decision: committee members added via the Members API always display as Member. Mapping to Manage/View is reserved for the Settings-updated path where Writer/Auditor roles have explicit semantics (per copilot-pull-request-reviewer).
- **`message_handler.go:661`** (goroutine closure) — Already safe: line 659 rebinds all used fields before the goroutine body runs, and Go 1.25 range-loop variables are per-iteration by definition (per copilot-pull-request-reviewer).

File: internal/service/email/committee_role_notification.go [hunk 20]
```diff
@@ -23,21 +25,87 @@ var (
...
 }
 
 // RenderCommitteeRoleNotification renders the subject, HTML body, and plain-text body
 // for a committee role notification email.
 func RenderCommitteeRoleNotification(data CommitteeRoleNotificationData) (subject, html, text string, err error) {
-	subject = sanitizeHeader(data.InviterName) + " added you as a " + data.Role + " on " + data.CommitteeName
+	if len(data.CapabilityGroups) == 0 {
+		data.CapabilityGroups = capabilityGroupsForRoles([]string{data.Role})
+	}
+	subject = sanitizeHeader(data.InviterName) + " added you as a " + data.Role + " on " + sanitizeHeader(data.CommitteeName)
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_role_notification.go]: Fixed in commit 2ceae7b — `data.Role` is now passed through `sanitizeHeader` in the subject line of `RenderCommitteeRoleNotification`, consistent with how `InviterName` and `CommitteeName` are already sanitized.

**Comment by andrest50** [comment]
## Review Feedback Addressed

Commit: 2ceae7b

### Changes Made
- **internal/service/email/committee_role_notification.go**: Sanitize `data.Role` via `sanitizeHeader` in the subject line of `RenderCommitteeRoleNotification`, closing the header-injection vector (per copilot-pull-request-reviewer)
- **committee_role_notification.html**, **committee_role_updated.html**: Wrapped `<ul style=...>` in `<!-- djlint:off H021 -->` / `<!-- djlint:on H021 -->` comments, matching the existing pattern used for the button anchor — email templates require inline styles for client compatibility (fixes MegaLinter djlint H021 failures)

### Threads Resolved
1 of 1 remaining unresolved threads addressed in this iteration.

**Review by dealako — APPROVED** [review]
Hi Andres! 👋

Great work on this PR. The notification lifecycle for committee role changes and member removal is well-architected — the classification logic in `classifyCommitteeUsers`, the display-role suppression via `effectiveRoleUnchanged`, and the `wasInvitedInOldSettings` guard all handle the tricky edge cases cleanly. Test coverage is thorough, including the no-op suppression matrix and the OldRoles-forwarding path through deletion. The multi-round responsiveness to CodeRabbit, Copilot, and CodeQL feedback has brought this to a solid state.

---

## Independent Review Findings

**Blocking:** 0  
**Minor:** 0  
**Nit:** 2

### Nits (no action required — both are already documented in-code)

- **`internal/service/message_handler.go` line 810 — `defer cancel()` inside goroutine:** The `context.WithTimeout` / `defer cancel()` pair lives inside the goroutine rather than outside it, which is fine here since the goroutine returns immediately after the `SendEmail` call and the context doesn't escape. The in-code comment explains this. Worth knowing for future readers who might reflexively move the defer out.

- **`internal/service/email/committee_role_notification.go` — `mapRoleToInviteRole` Manage default for unknown roles:** Defaulting to `InviteRoleManage` on unrecognized role names (with a `slog.Warn`) is a safe fallback, but it's a slightly privileged default. The warn log ensures it's visible in production. Already documented in the function comment.

---

## AI Comment Reconciliation

All bot findings are resolved or have documented rationale:

| Finding | Bot | Status | Commit |
|---|---|---|---|
| `make` capacity overflow in `classifyCommitteeUsers` | CodeQL | Fixed | `6d8dc3d` |
| Subject CR/LF injection on `CommitteeName` (4 threads) | Copilot | Fixed | `3ecf7cb` / `2ceae7b` |
| Stale comment on `newRoles` cardinality | Copilot | Fixed | `3ecf7cb` |
| Goroutine closure over range var | Copilot | Already correct — rebinding at line 659; author replied | — |
| Hard-coded `Role: "Member"` for member-created | Copilot | Intentional product decision; author replied | — |
| `effectiveRoleUnchanged` suppression vs PR description | Copilot | Intentional; PR description amended | — |
| `DOCKER_TAG := latest` regression | CodeRabbit | Fixed via `?= $(VERSION)` | `3ecf7cb` |
| Literal `"Member"` string vs invite-service constant | Copilot | Acknowledged; tracked in follow-up invite-service PR | — |
| Non-deterministic map iteration | Copilot | Fixed via `sort.Strings` | `e132906` |
| `wasInvitedInOldSettings` should require `Invite.UID != ""` | Copilot | Fixed | `3f9d8fc` |
| Test passes display names, not internal role names | Copilot | Author rejected with documented rationale | — |

---

## Revision Tracking (latest rounds)

- **`e132906`** — deterministic key ordering in `classifyCommitteeUsers` via `sort.Strings(sortedKeys)`
- **`2ceae7b`** — djlint H021 suppression on capability list `<ul>` + `sanitizeHeader` applied to `Role` in subject builder
- **`1c0f87e`** — role capabilities list added to notification email templates

---

## Security Review

Clean. No secrets, no injection vectors, no template HTML bypasses (`html/template` auto-escaping is preserved — no `template.HTML` casts on user data), `sanitizeHeader` correctly strips all ASCII control characters (including CR/LF) from all three subject components across all three notification types, and `redaction.Redact` is applied for PII in logs.

---

**Decision: ✅ Approved**

No blockers, no minor issues. The two nits are already documented in-code and require no changes. Branch is mergeable (no conflicts). Once the branch is rebased/merged up against `main`, this is ready to land.
